File: merkleroot.php
#!/usr/bin/php
<?php

// Swap Endian function
function swapendian($data) {
    return implode('', array_reverse(str_split($data, 2)));
}

// Merkle Root Function
function merklerootbinary($txids) {
 
	// Stop recursion if there is only one hash value left, because that's the merkle root.
	if (count($txids) == 1) {
		$merkleroot = $txids[0];
		return $merkleroot;
	}
 
	else {
 
		// Create the new array of hashes		
		while (count($txids) > 0) {
 
			if (count($txids) >= 2) {
				// Get first two
				$pair_first = $txids[0];
				$pair_second = $txids[1];
 
				// Hash them (double SHA256)
				$pair = $pair_first.$pair_second;
				$pairhashes[] = hash('sha256', hash('sha256', $pair, true), true);
 
				// Remove those two from the array
				unset($txids[0]);
				unset($txids[1]);
 
				// Re-set the indexes (the above just nullifies the values) and make a new array without the original first two slots.
				$txids = array_values($txids);
			}
 
			if (count($txids) == 1) {
				// Get the first one twice
				$pair_first = $txids[0];
				$pair_second = $txids[0];
 
				// Hash it with itself (double SHA256)
				$pair = $pair_first.$pair_second;
				$pairhashes[] = hash('sha256', hash('sha256', $pair, true), true);
 
				// Remove it from the array
				unset($txids[0]);
 
				// Re-set the indexes (the above just nullifies the values) and make a new array without the original first two slots.
				$txids = array_values($txids);
			}
 
		}
 
		// Recursion bit. Re-apply this function to the new array of hashes we've just created.
		return merklerootbinary($pairhashes);
 
	}
 
}
 
function merkleroot($txids) {
 
	// Convert txids in to big endian (BE), because that's the format they need to be in to get the merkle root.
	foreach ($txids as $txid) {
		$txidsBE[] = swapendian($txid);
	}
 
	// Now convert each of these txids in to binary, because the hash function wants the binary value, not the hex.
	foreach ($txidsBE as $txidBE) {
		$txidsBEbinary[] = hex2bin($txidBE);
	}
 
	// Work out the merkle root (in binary) using that lovely recursive function above.
	$merkleroot = merklerootbinary($txidsBEbinary);
 
	// Convert the merkle root in to hexadecimal and little-endian, because that's how it's stored in the block header.
	$merkleroot = swapendian(bin2hex($merkleroot));
 
	// Return it :)
	return $merkleroot;
 
}


// Get lines from STDIN
$lines = file('php://stdin');

// Clean up input (remove whitespace and ")
$txids = [];
foreach ($lines as $line) {
    $txids[] = str_replace('"', '', trim($line));
}

// Work out the Merkle Root
$merkleroot = merkleroot($txids);

// Print Result
echo $merkleroot."\n";

File: merkleroot.rb
require 'digest' # Need this for the SHA256 hash function

# Hash function used in the merkle root function (and in bitcoin in general)
def hash256(hex)
    binary = [hex].pack("H*")
    hash1 = Digest::SHA256.digest(binary)
    hash2 = Digest::SHA256.digest(hash1)
    result = hash2.unpack("H*")[0]
    return result
end

def merkleroot(txids)
  # 0. Keep an array of results for each level of hashing
  result = []

  # 1. Split up array in to pairs
  txids.each_slice(2) do |one, two|
    # 2. Concatenate each pair (or concatenate with itself if not a pair)
    if (two)
      concat = one + two
    else
      concat = one + one
    end

    # 3. Hash the concatenated pair and add to results array
    result << hash256(concat)
  end

  # Recursion: Exit Condition - Stop recursion when we have one final hash result.
  if result.length == 1
    # Convert the result to a string and return it
    return result.join('')
  end

  # Recursion: Do the same thing again for the array of hashed pairs.
  merkleroot(result)
end


# Test (e.g. block 000000000003ba27aa200b1cecaad478d2b00432346c3f1f3986da1afd33e506)
txids = [
  "8c14f0db3df150123e6f3dbbf30f8b955a8249b62ac1d1ff16284aefa3d06d87",
  "fff2525b8931402dd09222c50775608f75787bd2b87e56995a7bdd30f79702c4",
  "6359f0868171b1d194cbee1af2f16ea598ae8fad666d9b012c8ed2b79a236ec4",
  "e9a66845e05d5abc0ad04ec80f774a7e585c6e8db975962d069a522137b80c1d"
]

txids = txids.map {|x| x.scan(/../).reverse.join('') } # TXIDs must be in little endian
result = merkleroot(txids) # The result is in little endian, so lets convert it back to big endian...
puts result.scan(/../).reverse.join('') # f3e94742aca4b5ef85488dc37c06c3282295ffec960994b2c0d5ac2a25a95766
